storage/storage_image.go

```diff
@@ -93,309 +90,655 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
+		return nil
+	}
+	uncompressedLayerType := ""
+	switch manifestType {
+	case imgspecv1.MediaTypeImageManifest:
+		uncompressedLayerType = imgspecv1.MediaTypeImageLayer
+	case manifest.DockerV2Schema1MediaType, manifest.DockerV2Schema1SignedMediaType, manifest.DockerV2Schema2MediaType:
+		// This is actually a compressed type, but there's no uncompressed type defined
+		uncompressedLayerType = manifest.DockerV2Schema2LayerMediaType
+	}
+	for layerID != "" {
+		layer, err := s.imageRef.transport.store.Layer(layerID)
+		if err != nil {
+			logrus.Errorf("error reading layer %q in image %q: %v", layerID, s.ID, err)
+			return nil
+		}
+		if layer.UncompressedDigest == "" {
+			logrus.Errorf("uncompressed digest for layer %q is unknown", layerID)
+			return nil
+		}
+		if layer.UncompressedSize < 0 {
+			logrus.Errorf("uncompressed size for layer %q is unknown", layerID)
+			return nil
+		}
+		blobInfo := types.BlobInfo{
+			Digest:    layer.UncompressedDigest,
+			Size:      layer.UncompressedSize,
+			MediaType: uncompressedLayerType,
+		}
+		updatedBlobInfos = append([]types.BlobInfo{blobInfo}, updatedBlobInfos...)
+		layerID = layer.Parent
+	}
+	return updatedBlobInfos
+}
+
+// GetTargetManifest() is not supported.
+func (s *storageImageSource) GetTargetManifest(d digest.Digest) (manifestBlob []byte, MIMEType string, err error) {
+	return nil, "", ErrNoManifestLists
+}
+
+// GetSignatures() parses the image's signatures blob into a slice of byte slices.
+func (s *storageImageSource) GetSignatures(ctx context.Context) (signatures [][]byte, err error) {
+	var offset int
+	sigslice := [][]byte{}
+	signature := []byte{}
+	if len(s.SignatureSizes) > 0 {
+		signatureBlob, err := s.imageRef.transport.store.ImageBigData(s.ID, "signatures")
+		if err != nil {
+			return nil, errors.Wrapf(err, "error looking up signatures data for image %q", s.ID)
+		}
+		signature = signatureBlob
+	}
+	for _, length := range s.SignatureSizes {
+		sigslice = append(sigslice, signature[offset:offset+length])
+		offset += length
+	}
+	if offset != len(signature) {
+		return nil, errors.Errorf("signatures data contained %d extra bytes", len(signatures)-offset)
+	}
+	return sigslice, nil
+}
+
+// newImageDestination sets us up to write a new image, caching blobs in a temporary directory until
+// it's time to Commit() the image
 func newImageDestination(imageRef storageReference) (*storageImageDestination, error) {
+	directory, err := ioutil.TempDir(temporaryDirectoryForBigFiles, "storage")
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating a temporary directory")
+	}
+	// Break reading of the reference we're writing, so that copy.Image() won't try to rewrite
+	// schema1 image manifests to remove embedded references, since that changes the manifest's
+	// digest, and that makes the image unusable if we subsequently try to access it using a
+	// reference that mentions the no-longer-correct digest.
+	publicRef := imageRef
+	publicRef.name = nil
```

Comment:
`ImageDestination.Reference` is documented to record user intent, and stripping data like this is surprising; all the more when setting `.name` to `nil` does not change `StringWithinTransport`, i.e. there can be two different references with exactly equal string form but different behavior.

Could the `ImageDestination` instead get a direct ability to opt out of the `EmbeddedDockerReference` update, by adding a `NeedsEmbeddedDockerReferenceUpdate() bool` or something like that?

(FIXME? I guess this interacts with the `UpdatedLayerInfos` work which also aims to preserve the original manifest digest.)

Reply:
Yeah, the end-goal is to be able to read the layer blobs from the image when the digests of the blobs don't match the manifest, when changing the manifest would make the image inaccessible due to the `MatchesDigest` check failing.  `UpdatedLayerInfos` tries to address the former case, which affects images with all manifest types.  This hunk tries to avoid triggering an edit for stripping out an embedded docker reference, which mainly affects Schema1 images.